app/services/ingest_iati_activities.rb

```diff
@@ -0,0 +1,175 @@
+    if fund.nil?
+      puts "Could not find associated fund for: #{new_activity.previous_identifier} with `identifier`: #{programme_identifier}. Creating one." unless Rails.env.test?
+      fund = CreateFundActivity.new(organisation_id: service_owner_organisation.id).call
+      fund.update!(identifier: fund_identifier, wizard_status: :identifier)
+    end
+
+    programme = Activity.programmes.find_by(identifier: programme_identifier)
+    if programme.nil?
+      puts "Could not find associated programme for: #{new_activity.previous_identifier} with `identifier`: #{programme_identifier}. Creating one." unless Rails.env.test?
+      programme = CreateProgrammeActivity.new(organisation_id: service_owner_organisation.id, fund_id: fund.id).call
+      programme.update!(identifier: programme_identifier, wizard_status: :identifier)
+    end
+
+    programme
+  end
+
+  private def add_participating_organisation(delivery_partner:, new_activity:, legacy_activity:)
+    reporting_org_reference = legacy_activity.elements[1].attributes["ref"].value if legacy_activity.elements[1].name.eql?("reporting-org")
+    new_activity.reporting_organisation = Organisation.find_by(iati_reference: reporting_org_reference)
+
+    funding_org_element = legacy_activity.elements.detect { |element| element.name.eql?("participating-org") && element.attributes["role"].value.eql?("1") }
+    if funding_org_element
+      funding_org_reference = funding_org_element.attributes["ref"].value
+      funding_organisation = Organisation.find_by(iati_reference: funding_org_reference)
+      new_activity.funding_organisation_name = funding_organisation.name
+      new_activity.funding_organisation_type = funding_organisation.organisation_type
+      new_activity.funding_organisation_reference = funding_organisation.iati_reference
+    end
+
+    accountable_org_element = legacy_activity.elements.detect { |element| element.name.eql?("participating-org") && element.attributes["role"].value.eql?("2") }
+    if accountable_org_element
+      accountable_org_reference = accountable_org_element.attributes["ref"].value
+      accountable_organisation = Organisation.find_by(iati_reference: accountable_org_reference)
+      new_activity.accountable_organisation_name = accountable_organisation.name
+      new_activity.accountable_organisation_type = accountable_organisation.organisation_type
+      new_activity.accountable_organisation_reference = accountable_organisation.iati_reference
+    end
+
+    # There is no DP reference in the extending element. We must ask the user for it.
+    new_activity.extending_organisation = delivery_partner
+
+    implementing_org_elements = legacy_activity.elements.select { |element| element.name.eql?("participating-org") && element.attributes["role"].value.eql?("4") }
+    implementing_org_elements.each do |org_element|
+      new_activity.implementing_organisations << ImplementingOrganisation.create!(
```

Comment:
Create action are always called with a bang so the user is quickly notified of problems. This could be better built out with the ability to dry run this task and receive a digest of errors to fix all in one go.